packages/doenetml-worker-rust/lib-doenetml-core/src/core/general_state_var/string_state_var.rs

```diff
@@ -0,0 +1,158 @@
+    base_data_query: DataQuery,
+
+    default_value: String,
+}
+
+/// The data required to compute the value of this state variable.
+#[add_dependency_data]
+#[derive(Debug, Default, StateVariableDependencies, StateVariableDataQueries)]
+pub struct StringRequiredData {
+    /// If this state variable is extending another state variable,
+    /// the value that is being extended
+    extending: Option<StateVarView<String>>,
+
+    /// A vector of the string values of the dependencies coming from the base_data_query
+    base: Vec<StateVarView<String>>,
+
+    /// If not extending and don't have any base data,
+    /// then use this essential value, which will be initialized with the default value
+    essential: StateVarView<String>,
+}
+
+impl StringStateVar {
+    /// Creates a state var that queries its value from the given data query.
+    pub fn new(base_data_query: DataQuery) -> Self {
+        StringStateVar {
+            base_data_query,
+            ..Default::default()
+        }
+    }
+
+    /// Creates a state var that queries its value from children matching the `Text` profile.
+    pub fn new_from_children(default_value: String) -> Self {
+        StringStateVar {
+            base_data_query: DataQuery::Child {
+                match_profiles: vec![ComponentProfile::Text],
+                exclude_if_prefer_profiles: vec![],
+            },
+            default_value,
+        }
+    }
+
+    /// Creates a state var that queries its value from attr given by `attr_name`,
+    /// returning the attribute children that match the `Text` profile.
+    pub fn new_from_attribute(attr_name: AttributeName, default_value: String) -> Self {
+        StringStateVar {
+            base_data_query: DataQuery::AttributeChild {
+                attribute_name: attr_name,
+                match_profiles: vec![ComponentProfile::Text],
+            },
+            default_value,
+        }
+    }
+}
+
+impl StateVarUpdater<String, StringRequiredData> for StringStateVar {
+    fn default_value(&self) -> String {
+        self.default_value.clone()
+    }
+
+    fn return_data_queries(
+        &self,
+        extending: Option<ExtendSource>,
+        state_var_idx: StateVarIdx,
+    ) -> Vec<Option<DataQuery>> {
+        StringRequiredDataQueries {
+            extending: create_data_query_if_match_extend_source(extending, state_var_idx),
+            base: Some(self.base_data_query.clone()),
+            essential: Some(DataQuery::Essential),
+        }
+        .into()
+    }
+
+    fn calculate(data: &StringRequiredData) -> StateVarCalcResult<String> {
+        // concatenate the strings from extending and the base data query
+        let mut strings = if let Some(extending) = data.extending.as_ref() {
+            vec![extending]
+        } else {
+            vec![]
+        };
+
+        strings.extend(data.base.iter());
+
+        match strings.len() {
```

Comment:
I like the look of this :-)

Reply:
You can thank `clippy` for suggesting it.